File: apps/patch-library/src/sheet/fixturePatch/sceneryOptions.ts
import type { PatchedFixture } from "../../wire";
import { sceneryOf } from "./scenerySize";

export type SceneryOptions = NonNullable<PatchedFixture["scenery_options"]>;
export type ChainTopEnd = NonNullable<SceneryOptions["chain_top"]>;
export type ChainBottomEnd = NonNullable<SceneryOptions["chain_bottom"]>;
export type ChainMode = "plain" | "motor_top" | "motor_bottom";
export type SceneryOptionEdit = "scenery_colour" | "chain";

/** The patch sheet columns these choices are shown in, in table order. */
export const SCENERY_OPTION_COLUMNS = [
	"scenery_colour",
	"chain",
] as const;

/**
 * How a chain is rigged, and the end fittings each way stores. A hoist at one end hangs the chain
 * from a steelflex loop at the other; a plain chain is hooked directly at both.
 */
export const CHAIN_MODES: readonly {
	value: ChainMode;
	label: string;
	top: ChainTopEnd;
	bottom: ChainBottomEnd;
}[] = [
	{ value: "plain", label: "Plain chain", top: "direct", bottom: "direct" },
	{ value: "motor_top", label: "Motor on top", top: "motor", bottom: "steelflex_loop" },
	{
		value: "motor_bottom",
		label: "Motor on bottom",
		top: "steelflex_loop",
		bottom: "motor",
	},
];

const SRGB = /^#[0-9a-fA-F]{6}$/;

export function sceneryOptionsOf(fixture: PatchedFixture): SceneryOptions {
	return fixture.scenery_options ?? {};
}

export function isChain(fixture: PatchedFixture) {
	return sceneryOf(fixture)?.kind === "chain";
}

/**
 * How a chain is rigged, read from the hoist: one at the top or the bottom decides it, whatever the
 * other end stored. A chain placed before the choice existed has no top and hangs from a hoist.
 */
export function chainModeOf(fixture: PatchedFixture): ChainMode {
	const { chain_top, chain_bottom } = sceneryOptionsOf(fixture);
	if ((chain_top ?? "motor") === "motor") return "motor_top";
	return chain_bottom === "motor" ? "motor_bottom" : "plain";
}

export function chainModeLabel(mode: ChainMode) {
	return CHAIN_MODES.find((option) => option.value === mode)?.label ?? mode;
}

export function isSceneryOptionEdit(edit: string | null): edit is SceneryOptionEdit {
	return edit === "scenery_colour" || edit === "chain";
}

/**
 * The options after one choice, or why it is refused.
 *
 * An empty colour returns the object to its kind's own material, which is what it was drawn in
 * before anyone chose.
 */
export function sceneryOptionChange(
	fixture: PatchedFixture,
	edit: SceneryOptionEdit,
	value: string,
): { options: SceneryOptions } | { error: string } {
	const current = sceneryOptionsOf(fixture);
	if (edit === "scenery_colour") {
		const colour = value.trim();
		if (!colour) return { options: { ...current, colour_srgb: null } };
		if (!SRGB.test(colour))
			return { error: "Enter a colour as #RRGGBB, or clear it for the default." };
		return { options: { ...current, colour_srgb: colour.toUpperCase() } };
	}
	// Both ends are always written, so the stored fittings can never disagree with the mode.
	const mode = CHAIN_MODES.find((option) => option.value === value);
	return mode
		? { options: { ...current, chain_top: mode.top, chain_bottom: mode.bottom } }
		: { error: "Choose Plain chain, Motor on top or Motor on bottom." };
}

File: crates/shared/fixture/src/scenery_options.rs
//! What an operator chooses for one placed Venue object beyond its size.
//!
//! A curtain is bought in a colour and a chain is rigged with what hangs at each end. Neither is a
//! property of the product the profile describes — the same curtain profile is black serge on one
//! stage and white muslin on the next — so both travel with the placement, beside its size.

use serde::{Deserialize, Serialize};

/// Per-placement choices for a generated Venue object. Every field is optional: absent reads as
/// what the object's kind is drawn with by default, which is what every object placed before these
/// choices existed reads as.
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct SceneryOptions {
    /// The object's colour as `#RRGGBB` in sRGB. Absent keeps its kind's own material — black
    /// serge for a curtain, raw aluminium for truss.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub colour_srgb: Option<String>,
    /// What the top of a chain hangs from.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub chain_top: Option<ChainTopEnd>,
    /// What the bottom of a chain holds.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub chain_bottom: Option<ChainBottomEnd>,
}

impl SceneryOptions {
    pub fn is_empty(&self) -> bool {
        self.colour_srgb.is_none() && self.chain_top.is_none() && self.chain_bottom.is_none()
    }

    /// Whether every choice is one the object can be drawn with.
    pub fn validate(&self) -> Result<(), String> {
        if let Some(colour) = &self.colour_srgb
            && parse_srgb(colour).is_none()
        {
            return Err(format!(
                "scenery colour {colour:?} must be a #RRGGBB sRGB colour"
            ));
        }
        Ok(())
    }

    /// The chosen colour in linear light, for a renderer, when one is chosen.
    pub fn colour_linear(&self) -> Option<[f32; 3]> {
        let [red, green, blue] = parse_srgb(self.colour_srgb.as_deref()?)?;
        Some([
            srgb_to_linear(red),
            srgb_to_linear(green),
            srgb_to_linear(blue),
        ])
    }
}

/// The top of a chain: a hoist lifting it, or the chain made fast straight to the steel or truss.
#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ChainTopEnd {
    #[default]
    Motor,
    Direct,
}

/// The bottom of a chain: its hook shackled straight to the load, or a steelflex loop wrapped
/// around a beam or truss chord.
#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ChainBottomEnd {
    #[default]
    Direct,
    SteelflexLoop,
}

fn parse_srgb(value: &str) -> Option<[u8; 3]> {
    let hex = value.strip_prefix('#')?;
    if hex.len() != 6 || !hex.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return None;
    }
    let channel = |at: usize| u8::from_str_radix(&hex[at..at + 2], 16).ok();
    Some([channel(0)?, channel(2)?, channel(4)?])
}

fn srgb_to_linear(channel: u8) -> f32 {
    let value = f32::from(channel) / 255.0;
    if value <= 0.040_45 {
        value / 12.92
    } else {
        ((value + 0.055) / 1.055).powf(2.4)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn nothing_chosen_is_empty_and_serialises_to_nothing() {
        let options = SceneryOptions::default();
        assert!(options.is_empty());
        assert_eq!(
            serde_json::to_value(&options).unwrap(),
            serde_json::json!({})
        );
        let read: SceneryOptions = serde_json::from_value(serde_json::json!({})).unwrap();
        assert_eq!(read, options);
    }

    #[test]
    fn a_colour_is_a_hex_srgb_triple() {
        let red = SceneryOptions {
            colour_srgb: Some("#FF0000".into()),
            ..SceneryOptions::default()
        };
        assert!(red.validate().is_ok());
        assert_eq!(red.colour_linear(), Some([1.0, 0.0, 0.0]));
        for invalid in ["red", "#F00", "#GG0000", "FF0000"] {
            let options = SceneryOptions {
                colour_srgb: Some(invalid.into()),
                ..SceneryOptions::default()
            };
            assert!(options.validate().is_err(), "{invalid} was accepted");
        }
    }

    #[test]
    fn chain_ends_use_their_rigging_names() {
        let options = SceneryOptions {
            chain_top: Some(ChainTopEnd::Direct),
            chain_bottom: Some(ChainBottomEnd::SteelflexLoop),
            ..SceneryOptions::default()
        };
        assert_eq!(
            serde_json::to_value(&options).unwrap(),
            serde_json::json!({ "chain_top": "direct", "chain_bottom": "steelflex_loop" })
        );
    }
}
